enhancements/20240805-encrypted-block-volume.md

```diff
@@ -0,0 +1,335 @@
++	SourceParameters VirtualMachineImageSourceParameters `json:"sourceParameters"`
+}
+
++type VirtualMachineImageSourceParameters struct {
++	// +optional
++	SecretName string `json:"secretName"`
++
++	// +optional
++	SecretNamespace string `json:"secretNamespace"`
+
++	// +optional
++	// +kubebuilder:validation:Enum=encrypt;decrypt
++	Encryption VirtualMachineImageEncryptionType `json:"encryption"`
+
++	// +optional
++	SourceVirtualMachineImageName string `json:"sourceVirtualMachineImageName"`
+
++	// +optional
++	SourceVirtualMachineImageNamespace string `json:"sourceVirtualMachineImageNamespace"`
++}
+```
+
+### Design
+
+
+### Implementation Overview
+
+Harvester uses similar parameters in Longhorn, so we just need to sync those parameters to backing image in Longhorn to activate the encryption/decryption feature.
+
+For example:
+
+Let's say we have this:
+
+```yaml
+apiVersion: harvesterhci.io/v1beta1
+kind: VirtualMachineImage
+metadata:
+    name: encrypted-alpine-extended
+    namespace: default
+spec:
+    displayName: encrypted-alpine-extended-3.20.2-x86_64.iso
+    retry: 3
+    sourceType: clone
+    storageClassParameters:
+        migratable: "true"
+        numberOfReplicas: "1"
+        staleReplicaTimeout: "30"
+    sourceParameters:
+        secretName: "longhorn-crypto"
+        secretNamespace: "longhorn-system"
+        encryption: "encrypt"
+        sourceVirtualMachineImageName: "alpine-extended" # not display name
+        sourceVirtualMachineImageNamespace: "default"
+```
+
+The `sourceVirtualMachineImageName` is from here:
+
+```yaml
+apiVersion: harvesterhci.io/v1beta1
+kind: VirtualMachineImage
+metadata:
+  name: alpine-extended
+  namespace: default
+spec:
+  displayName: alpine-extended-3.20.2-x86_64.iso
+  retry: 3
+  sourceType: download
+  storageClassParameters:
+    migratable: "true"
+    numberOfReplicas: "3"
+    staleReplicaTimeout: "30"
+  url: https://dl-cdn.alpinelinux.org/alpine/v3.20/releases/x86_64/alpine-extended-3.20.2-x86_64.iso
+```
+
+
+We'll sync those parameters to backing image in Longhorn.
+
+```yaml
+apiVersion: longhorn.io/v1beta2
+kind: BackingImage
+metadata:
+   name: default-encrypted-alpine-extended
+   namespace: longhorn-system
+spec:
+   minNumberOfCopies: 3
+   sourceParameters:
+     backing-image: "{this is the backing image name, will get it via VirtualMachineImage}"
+     encryption: "encrypt"
+     secret: "longhorn-crypto"
+     secret-namespace: "longhorn-system"
+   sourceType: clone
+```
+
+Then, we could get the encrypted backing image in Longhorn, and use it as boot volume in Harvester.
+
+
+### Test plan
+
+- Verify that volumes can be encrypted and decrypted as specified. 
+- Verify that encrypted boot volume and data volume can be attached to virtual machines.
+- Verify that the encrypted data volume, when used as a hot plug volume, can be attached to virtual machines.
+- Ensure that encrypted volumes are not accessible without the correct decryption keys. 
```

Comment:
Do you have details abut this?

Can user create the decryption key when needed instead of initially.

Reply:
The logic behind the encryption/decryption uses LUKS. If you know the key, you can still decrypt volume to read the data on the node.